### DISCUSSION TITLE
Vercel & Custom API Backend

### DISCUSSION THREAD

**Discussion by Djboy08** [vercel/community#510] [Help]
Hello! 

I am doing work for a client and they want to use a Custom API Backend, but also keep Vercel as the front end. Now for the most part this is easy to do. The troubling part I find from this is when it comes the sessions and authentication. 

Ill be using discord OAuth as auth, while using next-auth to handle it. However, the one issue I keep finding myself in is the vercel backend vs the custom api backend. 

There are some cases where I want to use SSR on a page with authenticated data. However, if the session is located on the Custom API Backend and not the vercel backend, then I won’t be able to confirm authentication and get the session data. 

It basically makes SSR obsolete if it requires auth data since the vercel back-end has no way of getting that session data. So what would be the solution to this? Ive read online that you can share sessions between two servers but then you would need some database to make sure they share the same info, which I really don’t want to do.

Any other solutions? 

Thanks!

**Comment by ShawnCentra** [comment]
My company uses a custom CMS internally. 

Custom API + Vercel doesn't work since the API route has to be made public. Vercel is a hosting service and therefore requires you to have proper settings for your custom API to communicate with it - this is what i've learned. 

so if you  cant access customAPI.com/userToken=1234 publically, you can't use Vercel to get that data.

**Reply by Djboy08** [reply]
The backend API would be public, we mainly wanted a custom backend due to more control over rate limiting and managing stuff like that.

**Reply by ShawnCentra** [reply]
Thats the limiting factor if its public and reachable or not. It should be okay to implement with Vercel now using RESTful consumption for the API. 

However your issue is with persistent session data between two servers, which you've found a solution for (database for storing the data between the custom API and then consuming it via vercel backend)

**Reply by Djboy08** [reply]
Is that the only route I can take? Is there any other options than to not create a database shared between them?

**Reply by ShawnCentra** [reply]
I would explore Oauth implementation into the Custom API so that when you fetch session data from custom api, Oauth has that data.

but if the flow doesnt match up here - then yea you might have to sync a DB. Otherwise how will you get that data from server 1 -> server 2?  

thats what I'd do. but maybe someone has a better solution? I just dont see one myself.

**Reply by ShawnCentra** [reply]
also check other similar questions - feels like this is a common topic 

https://stackoverflow.com/questions/17854156/oauth2-for-custom-api
https://stackoverflow.com/questions/63194644/oauth-login-with-custom-api

**Reply by Djboy08** [reply]
If I had the following Workflow

	Vercel Backend
	Vercel Frontend
	Custom API Backend

and if I use next-auth on the Custom API Backend side, then I will have clear sessions from the the vercel frontend to the Custom API Backend.


Now the issue is making Vercel Backend make use of our custom API Domain. 

On the API routes for the vercel backend, if there is user data we want, we will send the request to our Custom API Domain including the cookie in the header or something that next-auth uses. 

On our Custom API Domain we would have a `/api/auth.js` for example, this will accept a cookie that is used to look up that the cookie exists and return data on it. 

For solving the ServerSidedProps issue, we could do the same thing where we request the Custom API Backend with a cookie from in the header to tell us if they are authenticated or not, and if they are who they are.


Would this function? Or am I getting ahead of myself and forgetting something?

**Reply by Djboy08** [reply]
basically vercel backend would act as a proxy in some cases, specifically in SSR
